File: api-python/app/messages/MessageFactory.py
from typing import Type, Dict, Any
from .UnknowMessage import UnknowMessage
from .MessageType import MessageType
from .QuestionMessage import QuestionMessage
from .AnswerMessage import AnswerMessage
from .StatusMessage import StatusMessage
from .EndMessage import EndMessage
from .BaseMessage import BaseMessage

# Mapeo entre MessageType y su clase correspondiente
message_type_map: Dict[MessageType, Type[BaseMessage]] = {
    MessageType.QUESTION: QuestionMessage,
    MessageType.ANSWER: AnswerMessage,
    MessageType.STATUS: StatusMessage,
    MessageType.END: EndMessage
}

def MessageFactory(data: dict) -> BaseMessage:
    """
    Fabrica un mensaje a partir de los datos proporcionados.

    Esta función construye un objeto de tipo `BaseMessage` a partir de un diccionario que contiene
    los datos del mensaje. El tipo de mensaje se determina a partir del campo "type" en los datos,
    y se mapea al tipo correspondiente utilizando el diccionario `message_type_map`.

    Si el tipo de mensaje no es reconocido o está ausente, se retorna un `UnknowMessage`.

    Args:
        data (dict): Diccionario con los datos del mensaje, incluyendo el tipo y el contenido.

    Returns:
        BaseMessage: El mensaje correspondiente al tipo indicado en `data`, o un `UnknowMessage` si el tipo es inválido.
    """
    try:
        message_type = MessageType(data["type"])
    except ValueError as e:
        return UnknowMessage(data)

    message_class = message_type_map.get(message_type)
    if not message_class:
        return UnknowMessage(data)

    return message_class(**data)

File: frontend/src/models/messages/MessageFactory.ts
import { MessageType } from "./MessageType";
import QuestionMessage from "./QuestionMessage";
import MessageInterface from "./MessageInterface";

import AnswerMessage from "./AnswerMessage";
import UnknowMessage from "./UnknowMessage";
import StatusMessage from "./StatusMessage";
import EndMessage from "./EndMessage";
import SocketMessageInterface from "./SocketMessageInterface";
import ErrorMessage from "./ErrorMessage";
import { ErrorMessageData } from "./ErrorMessageData";

/**
 * Fábrica de mensajes que crea instancias de diferentes tipos de mensajes basados en el tipo recibido.
 */
export default class MessageFactory {

    /**
     * Crea una instancia de un mensaje basado en la información recibida.
     * @param message - Objeto de tipo SocketMessageInterface que contiene los datos del mensaje.
     * @returns Una instancia de MessageInterface si el tipo de mensaje es válido, de lo contrario, retorna null.
     */
    static create(message: SocketMessageInterface): MessageInterface | null {
        switch (message.type) {
            case MessageType.QUESTION:
                return new QuestionMessage(message.id, message.content);
            
            case MessageType.ANSWER:
                return new AnswerMessage(message.id, message.content);

            case MessageType.STATUS:
                return new StatusMessage(message.id, message.content);

            case MessageType.END:
                    return new EndMessage(message.id);

            case MessageType.ERROR:
                    return new ErrorMessage(message.content);
            
            default:
                return new UnknowMessage();
        }
    }
}
